### PR TITLE
chaosctl add recover subcommand

### PR DESCRIPTION
**Pull request by Hexilee**
### What problem does this PR solve?

Add recover subcommand for chaosctl. This is a part split from #2992.

### Related changes

- [ ] Need to update `chaos-mesh/website`
- [ ] Need to update `Dashboard UI`
- Need to **cheery-pick to release branches**
  - [ ] release-2.1
  - [ ] release-2.0

### Checklist

Tests

- [ ] Unit test
- [ ] E2E test
- [ ] No code
- [ ] Manual test (add steps below)

Side effects

- [ ] Breaking backward compatibility

### Release note 

```text
Please add a release note.

You can safely ignore this section if you don't think this PR needs a release note.
```

### DCO

If you find the DCO check fails, please run commands like below (Depends on the actual situations. For example, if the failed commit isn't the most recent) to fix it:

```shell
git commit --amend --signoff
git push --force
```

### CODE REVIEW

File: pkg/chaosctl/cmd/recover.go [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+		}
+	}
+	return nil
+}
+
+// List pods to recover
+func (o *RecoverOptions) List(client *ctrlclient.CtrlClient) ([]string, cobra.ShellCompDirective) {
+	pods, err := o.selectPods(client, []string{})
+	if err != nil {
+		common.PrettyPrint(errors.Wrap(err, "select pods").Error(), 0, common.Red)
```
Comment (WangXiangUSTC) [inline pkg/chaosctl/cmd/recover.go:142]: can return here

File: pkg/chaosctl/cmd/recover.go [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+	var names []string
+	for _, pod := range pods {
+		names = append(names, pod.Name)
+	}
+
+	return names, cobra.ShellCompDirectiveNoFileComp
+}
+
+func (o *RecoverOptions) selectPods(client *ctrlclient.CtrlClient, names []string) ([]*ctrlclient.PartialPod, error) {
```
Comment (WangXiangUSTC) [inline pkg/chaosctl/cmd/recover.go:153]: the argument `names` is useless

File: pkg/chaosctl/cmd/recover.go [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+		GenericSelectorSpec: v1alpha1.GenericSelectorSpec{
+			Namespaces: []string{o.namespace},
+		},
+	}
+
+	labelSelector := map[string]string{}
+	if o.labels != nil && len(*o.labels) > 0 {
+		for _, label := range *o.labels {
+			pair := strings.Split(label, "=")
+			if len(pair) == 2 {
```
Comment (WangXiangUSTC) [inline pkg/chaosctl/cmd/recover.go:167]: if not, the label is invalid,  should we print a log?

File: pkg/chaosctl/cmd/recover.go (hunk 1 above)
Comment (Hexilee) [reply pkg/chaosctl/cmd/recover.go:142]: Fixed

File: pkg/chaosctl/cmd/recover.go (hunk 2 above)
Comment (Hexilee) [reply pkg/chaosctl/cmd/recover.go:154]: Used

File: pkg/chaosctl/cmd/recover.go (hunk 3 above)
Comment (Hexilee) [reply pkg/chaosctl/cmd/recover.go]: Sure

File: pkg/chaosctl/cmd/recover.go (hunk 2 above)
Comment (WangXiangUSTC) [reply pkg/chaosctl/cmd/recover.go:154]: oh, my mistake 🥲

File: pkg/chaosctl/cmd/recover.go [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+		Short: `Recover certain chaos from certain pods`,
+		Long: `Recover certain chaos from certain pods.
+Currently unimplemented.
+
+Examples:
+  # Recover network chaos from pods in namespace default
+  chaosctl recover networkchaos
+
+  # Recover network chaos from certain pod in 
+  chaosctl recover networkchaos PODs -n NAMESPACE`,
```
Comment (WangXiangUSTC) [inline pkg/chaosctl/cmd/recover.go:52]: how about adding an example of how to set the PODs, seems that it is a array

**Review by WangXiangUSTC — APPROVED** [review]
rest LGTM

File: pkg/chaosctl/cmd/recover.go (hunk 4 above)
Comment (Hexilee) [reply pkg/chaosctl/cmd/recover.go]: Done

**Review by WangXiangUSTC — APPROVED** [review]
LGTM

File: pkg/chaosctl/cmd/recover.go [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+	o := &RecoverOptions{namespace: "default"}
+
+	recoverCmd := &cobra.Command{
+		Use:   `recover (CHAOSTYPE) (PODs) [-n NAMESPACE]`,
+		Short: `Recover certain chaos from certain pods`,
+		Long: `Recover certain chaos from certain pods.
+Currently unimplemented.
+
+Examples:
+  # Recover network chaos from pods in namespace default
```
[the remark this answers is not in the document]
Comment (Hexilee) [reply pkg/chaosctl/cmd/recover.go:48]: Doesn't support it, it's designed to forcedly recover pods, we should not assume the Chaos Mesh still works as expected. Moreover, if multiple experiments affect the same pod, we cannot only recover one of them forcedly.

File: pkg/chaosctl/cmd/recover.go [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+		ValidArgsFunction: noCompletions,
+	}
+
+	for chaosType, builder := range builders {
+		recoverCmd.AddCommand(recoverResourceCommand(o, chaosType, builder))
+	}
+
+	recoverCmd.PersistentFlags().StringVarP(&o.namespace, "namespace", "n", "default", "namespace to find pods")
+	o.labels = recoverCmd.PersistentFlags().StringSliceP("label", "l", nil, "labels to select pods")
+	err := recoverCmd.RegisterFlagCompletionFunc("namespace", func(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
```
[the remark this answers is not in the document]
Comment (Hexilee) [reply pkg/chaosctl/cmd/recover.go:65]: When you enter TAB (auto-completion).

File: pkg/chaosctl/debug/networkchaos.go [hunk 7]
```diff
@@ -54,8 +55,8 @@ func (d *networkDebugger) Collect(ctx context.Context, namespace, chaosName stri
 					Name      string
 					Pod       struct {
 						Ipset    string
```
[the remark this answers is not in the document]
Comment (Hexilee) [reply pkg/chaosctl/debug/networkchaos.go:57]: IPSets are unnecessary in forced recovery, we don't need to parse it server-side in this PR. If it's necessary for other use-cases, we should parse it in another PR.

File: pkg/chaosctl/cmd/recover.go [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+		GenericSelectorSpec: v1alpha1.GenericSelectorSpec{
+			Namespaces: []string{o.namespace},
+		},
+	}
+
+	if len(names) != 0 {
+		selector.Pods = map[string][]string{o.namespace: names}
+	}
+
+	labelSelector := map[string]string{}
```
[the remark this answers is not in the document]
Comment (Hexilee) [reply pkg/chaosctl/cmd/recover.go]: Done

File: pkg/ctrl/client/ctrlclient.go [hunk 9]
```diff
@@ -19,13 +19,27 @@ import (
...
+	"github.com/chaos-mesh/chaos-mesh/api/v1alpha1"
+	"github.com/chaos-mesh/chaos-mesh/pkg/ctrl/server/model"
 )
 
 type CtrlClient struct {
 	QueryClient        *graphql.Client
 	SubscriptionClient *graphql.SubscriptionClient
 }
 
+type PartialPod struct {
```
[the remark this answers is not in the document]
Comment (Hexilee) [reply pkg/ctrl/client/ctrlclient.go]: Sure, now the `PartialPod ` is moved to `pkg/chaosctl/recover` .

File: pkg/chaosctl/cmd/recover.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (Andrewmatilde) [reply pkg/chaosctl/cmd/recover.go:49]: Why just default namespace here? Can we support recover on a specific namespace?

File: pkg/chaosctl/cmd/recover.go [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+type RecoverOptions struct {
+	namespace string
+	labels    *[]string
+}
+
+func NewRecoverCommand(logger logr.Logger, builders map[string]recover.RecoverBuilder) (*cobra.Command, error) {
+	o := &RecoverOptions{namespace: "default"}
+
+	recoverCmd := &cobra.Command{
+		Use:   `recover (CHAOSTYPE) (PODs) [-n NAMESPACE]`,
```
Comment (Andrewmatilde) [inline pkg/chaosctl/cmd/recover.go:43]: It is rare to use (...) in this situation, for `PODS` you can provide `pod[,pod[,pod...]]`.

File: pkg/chaosctl/cmd/recover.go [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "register completion func for flag `label`")
+	}
+	return recoverCmd, nil
+}
+
+func recoverResourceCommand(option *RecoverOptions, chaosType string, builder recover.RecoverBuilder) *cobra.Command {
+	return &cobra.Command{
+		Use:   fmt.Sprintf(`%s (PODs) [-n NAMESPACE]`, chaosType),
```
Comment (Andrewmatilde) [inline pkg/chaosctl/cmd/recover.go:96]: ditto

**Review by Andrewmatilde — COMMENTED** [review]
Rest LGTM, I think some of this PR may be changed in the following PR of recover in chaosctl.

File: pkg/chaosctl/cmd/recover.go (hunk 10 above)
Comment (Hexilee) [reply pkg/chaosctl/cmd/recover.go]: Reasonable, modified.

File: pkg/chaosctl/cmd/recover.go (hunk 11 above)
Comment (Hexilee) [reply pkg/chaosctl/cmd/recover.go]: Done

**Comment by Hexilee** [comment]
@Andrewmatilde PTAL again!

**Review by cwen0 — APPROVED** [review]
LGTM

**Pull request merged**